### ISSUE TITLE
Cannot download file from HTTPS server

### ISSUE THREAD

**Issue by ealymbaev** [android-async-http/android-async-http#1029]
I am trying to download the image from the url, but it gives me SSL Handshake error (see below).

There is the code I am using for downloading (the url is also there):

```
String url = "https://cdn.grouvi.im/uploads/d4fec456-8ca6-49f8-b141-3a3a0c4ea85d/tmp075dc37c06c343f3adc69c0c07bd42eb.jpg";

AsyncHttpClient client = new AsyncHttpClient();

client.get(url, new FileAsyncHttpResponseHandler(App.getContext()) {
    @Override
    public void onFailure(int statusCode, Header[] headers, Throwable throwable, File file) {
        Log.e("LOOPJ", "FAILURE " + statusCode);
        throwable.printStackTrace();
    }

    @Override
    public void onSuccess(int statusCode, Header[] headers, File response) {
        Log.e("LOOPJ", "SUCCESS " + statusCode);
    }
});
```

I have also tried to download the file using the native Android way:

```
new AsyncTask<Void, Void, Void>() {
    @Override
    protected Void doInBackground(Void... voids) {
        try {
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setDoInput(true);

            conn.connect();
            int response = conn.getResponseCode();
            Log.e("GRV", "The response is: " + response);
            InputStream is = conn.getInputStream();

            String contentAsString = readIt(is, 500);

            Log.e("GRV", "Result: " + contentAsString);

        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }
}.execute();
```

And IT WORKED!

But the loopj lib cannot download it, giving me the following error:

`11-23 11:32:41.330 29924-31935/im.grouvi.app.debug W/System.err: javax.net.ssl.SSLHandshakeException: Handshake failed`

`11-23 11:32:41.330 29924-31935/im.grouvi.app.debug W/System.err: Caused by: javax.net.ssl.SSLProtocolException: SSL handshake terminated: ssl=0x9791f1c0: Failure in SSL library, usually a protocol error
11-23 11:32:41.330 29924-31935/im.grouvi.app.debug W/System.err: error:100c5410:SSL routines:ssl3_read_bytes:SSLV3_ALERT_HANDSHAKE_FAILURE (external/boringssl/src/ssl/s3_pkt.c:972 0x9375a480:0x00000001)
11-23 11:32:41.330 29924-31935/im.grouvi.app.debug W/System.err: error:100c009f:SSL routines:ssl3_get_server_hello:HANDSHAKE_FAILURE_ON_CLIENT_HELLO (external/boringssl/src/ssl/s3_clnt.c:750 0xab1b950f:0x00000000)`

Could you please resolve this issue as I cannot use the library for development anymore because of this issue.

**Comment by SadieYuCN** [comment]
I think you should try to get AsyncHttpClient instance by new AsyncHttpClient(true, 80, 443) , if you want to get something from https

**Comment by ealymbaev** [comment]
@SadieYuCN I have tried it, sorry for not describing it initially. But it does not affect anything. The errors are still present.

**Comment by ealymbaev** [comment]
Any solution for my issue? Need it urgently

**Comment by mcamocci** [comment]
i removed 's'  from https to htttp, and things worked for me.

**Comment by dev01-adroitinfosol** [comment]
Here is the solution for you, This can be applied when you are using server for development purpose.

public SSLContext getSslContext() {

    TrustManager[] byPassTrustManagers = new TrustManager[] { new X509TrustManager() {
        public X509Certificate[] getAcceptedIssuers() {
            return new X509Certificate[0];
        }

        public void checkClientTrusted(X509Certificate[] chain, String authType) {
        }

        public void checkServerTrusted(X509Certificate[] chain, String authType) {
        }
    } };

    SSLContext sslContext=null;

    try {
        sslContext = SSLContext.getInstance("TLS");
    } catch (NoSuchAlgorithmException e) {
        e.printStackTrace();
    }
    try {
        sslContext.init(null, byPassTrustManagers, new SecureRandom());
    } catch (KeyManagementException e) {
        e.printStackTrace();
    }

    return sslContext;

}

 AsyncHttpClient client = new AsyncHttpClient();

 client.setSSLSocketFactory(
            new SSLSocketFactory(getSslContext(),
            SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER));

**Comment by ashishtikarye** [comment]
**please post any solutions i tried above solutions also but still issue not resolved.**
**Error during handshake api from android with android-async-http:1.4.9**
`Unknown error during handshake javax.net.ssl.SSLException: SSL handshake aborted: ssl=0x61b78a98: I/O error during system call, Connection reset by peer at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method) at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:406) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:543) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:409) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:479) at cz.msebera.android.httpclient.conn.scheme.SchemeSocketFactoryAdaptor.connectSocket(SchemeSocketFactoryAdaptor.java:66) at cz.msebera.android.httpclient.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:177) at cz.msebera.android.httpclient.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:145) at cz.msebera.android.httpclient.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:131) at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:611) at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:446) at cz.msebera.android.httpclient.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:879) at cz.msebera.android.httpclient.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82) at com.loopj.android.http.AsyncHttpRequest.makeRequest(AsyncHttpRequest.java:146) at com.loopj.android.http.AsyncHttpRequest.makeRequestWithRetries(AsyncHttpRequest.java:177) at com.loopj.android.http.AsyncHttpRequest.run(AsyncHttpRequest.java:106) at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422) at java.util.concurrent.FutureTask.run(FutureTask.java:237) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) at java.lang.Thread.run(Thread.java:841) javax.net.ssl.SSLException: SSL handshake aborted: ssl=0x61b78a98: I/O error during system call, Connection reset by peer at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method) at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:406) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:543) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:409) at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:479) at cz.msebera.android.httpclient.conn.scheme.SchemeSocketFactoryAdaptor.connectSocket(SchemeSocketFactoryAdaptor.java:66) at cz.msebera.android.httpclient.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:177) at cz.msebera.android.httpclient.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:145) at cz.msebera.android.httpclient.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:131) at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:611) at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:446) at cz.msebera.android.httpclient.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:879) at cz.msebera.android.httpclient.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82) at com.loopj.android.http.AsyncHttpRequest.makeRequest(AsyncHttpRequest.java:146) at com.loopj.android.http.AsyncHttpRequest.makeRequestWithRetries(AsyncHttpRequest.java:177) at com.loopj.android.http.AsyncHttpRequest.run(AsyncHttpRequest.java:106) at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422) at java.util.concurrent.FutureTask.run(FutureTask.java:237) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)`

**Comment by Zapnologica** [comment]
We are also currently experiencing this issue, We have migrated out API to azure web app which does not support old TLS version, and all users with older phones android 6 below seem to be moaning about getting instant networking error.

**Comment by tiaan-lg** [comment]
The Azure web app uses Azure Traffic manager to route to the endpoint, and the endpoint has valid SSL cert using TLS version 1. If we route directly to the app it works but going through the Traffic manager fails with a handshake error. The traffic manager uses Server Name Indication (SNI) to present certificates this is what i suspect is causing the problem

**Comment by shwuhk** [comment]
@smarek Any idea to fix this? Thanks!

**Comment by saeed-rz** [comment]
fix on 1.4.11, try to implement 
`implementation 'org.conscrypt:conscrypt-android:2.4.0'`
then call this method before creating new object from AsyncHttpClient
`AsyncHttpClient.useConscryptSSLProvider();`
